network/dag/bbolt_tree.go

```diff
@@ -1,38 +1,143 @@
 )
 
+// treeBucketFillPercent can be much higher than default 0.5 since the keys (unique node ids)
+// should be added to the bucket in monotonic increasing order, and the values are of fixed size.
+const treeBucketFillPercent = 0.9
+const defaultObserverRollbackTimeOut = 10 * time.Second
+
+var observerRollbackTimeOut = defaultObserverRollbackTimeOut
+
 type bboltTree struct {
-	db *bbolt.DB
+	db                *bbolt.DB
+	bucketFillPercent float64
+	bucketName        string
+	tree              tree.Tree
+}
+
+// newBBoltTreeStore returns an instance of a BBolt based tree store. Buckets managed by this store are filled to treeBucketFillPercent
+func newBBoltTreeStore(db *bbolt.DB, bucketName string, tree tree.Tree) *bboltTree {
+	return &bboltTree{
+		db:                db,
+		bucketFillPercent: treeBucketFillPercent,
+		bucketName:        bucketName,
+		tree:              tree,
+	}
 }
 
-// newBBoltTreeStore returns an instance of a BBolt based tree store. Buckets managed by this store are filled to BucketFillPercent
-func newBBoltTreeStore(db *bbolt.DB) *bboltTree {
-	return &bboltTree{db: db}
+// getRoot returns the tree.Data summary of the entire tree.
+func (store *bboltTree) getRoot() tree.Data {
+	return store.tree.GetRoot()
+}
+
+// getZeroTo returns the tree.Data sum of all tree pages/leaves upto and including the one containing the requested Lamport Clock value.
+// In addition to the data, the highest LC value of this range is returned.
+func (store *bboltTree) getZeroTo(clock uint32) (tree.Data, uint32) {
+	return store.tree.GetZeroTo(clock)
+}
+
+func (store *bboltTree) isEmpty() bool {
+	return store.tree.GetRoot().IsEmpty()
+}
+
+// dagCallback inserts a transaction reference to the in-memory tree and to persistent storage.
+// The tree is not aware of previously seen transactions, so it should be transactional with updates to the dag.
+func (store *bboltTree) dagObserver(ctx context.Context, transaction Transaction, _ []byte) {
+	if transaction != nil { // can happen when payload is written for private TX
+		err := storage.BBoltTXUpdate(ctx, store.db, func(callbackCtx context.Context, tx *bbolt.Tx) error {
+			dirty := store.tree.InsertGetDirty(transaction.Ref(), transaction.Clock())
+
+			// Rollback after timeout to bring tree and DAG back in sync.
+			// A call to writeUpdates will persist all uncommitted tree changes. So a failed bboltTx will be dropped by the dag and (eventually) persisted by the tree.
+			c, cancel := context.WithTimeout(context.Background(), observerRollbackTimeOut) // << timeout must not be shorter than expected write operation to disk
+			go func() {
+				<-c.Done()
+				err := c.Err()
+				if err == context.DeadlineExceeded {
+					log.Logger().Warnf("deadline exceeded - rollback transaction %s from %s", transaction.Ref(), store.bucketName)
+					store.tree.Delete(transaction.Ref(), transaction.Clock())
+				}
+			}()
+			tx.OnCommit(func() {
+				store.tree.ResetUpdate()
+				cancel()
```

Comment:
The cancel() will Delete() the transaction from the tree and (eventually from) its storage, so the tree will not be corrupt. 

However, the tree will now be out of sync with the dag. Since the tree is used for set reconciliation, the node should request the 'missing' transaction from its peers, but the Tx will be dropped (not offered to the dag observers again) since it is already in the dag. 

It is even possible that both trees and the dag in `state` all have a different state.

Any suggestions on what to do with this?
- We could add metrics to log this?
- How likely is this to happen with the current timeout of 10 sec?

any thoughts on this @woutslakhorst